File: src/main/java/com/budiyev/android/imageloader/InternalUtils.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Yuriy Budiyev [yuriy.budiyev@yandex.ru]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.budiyev.android.imageloader;

import java.io.ByteArrayInputStream;
import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;

import android.content.ContentResolver;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.media.ExifInterface;
import android.view.View;
import android.widget.ImageView;

final class InternalUtils {
    private static final int CONNECT_TIMEOUT = 10000;
    private static final int MAX_POOL_SIZE = 4;
    private static final int MIN_POOL_SIZE = 1;
    private static final String URI_SCHEME_HTTP = "http";
    private static final String URI_SCHEME_HTTPS = "https";
    private static final String URI_SCHEME_FTP = "ftp";

    private InternalUtils() {
    }

    public static void invalidate(@Nullable final ImageCache memoryCache, @Nullable final ImageCache storageCache,
            @NonNull final DataDescriptor<?> descriptor) {
        final String key = descriptor.getKey();
        if (key == null) {
            return;
        }
        if (memoryCache != null) {
            memoryCache.remove(key);
        }
        if (storageCache != null) {
            storageCache.remove(key);
        }
    }

    @Nullable
    public static String buildFullKey(@Nullable final String base, @Nullable final Size requiredSize,
            @Nullable final BitmapTransformation transformation) {
        if (base == null) {
            return null;
        }
        if (requiredSize == null && transformation == null) {
            return base;
        }
        final StringBuilder sb = new StringBuilder(base);
        if (requiredSize != null) {
            sb.append("_required_size_").append(requiredSize.getWidth()).append("x").append(requiredSize.getHeight());
        }
        if (transformation != null) {
            sb.append(transformation.getKey());
        }
        return sb.toString();
    }

    public static void close(@Nullable final Closeable closeable) {
        if (closeable == null) {
            return;
        }
        try {
            closeable.close();
        } catch (final IOException ignored) {
        }
    }

    @Nullable
    public static InputStream getDataStreamFromUri(@NonNull final Context context, @NonNull final Uri uri)
            throws IOException {
        final String scheme = uri.getScheme();
        if (URI_SCHEME_HTTP.equalsIgnoreCase(scheme) || URI_SCHEME_HTTPS.equalsIgnoreCase(scheme) ||
                URI_SCHEME_FTP.equalsIgnoreCase(scheme)) {
            final URLConnection connection = new URL(uri.toString()).openConnection();
            connection.setConnectTimeout(CONNECT_TIMEOUT);
            return connection.getInputStream();
        } else {
            return context.getContentResolver().openInputStream(uri);
        }
    }

    @Nullable
    public static InputStream getDataStreamFromUrl(@NonNull final String url) throws IOException {
        final URLConnection connection = new URL(url).openConnection();
        connection.setConnectTimeout(CONNECT_TIMEOUT);
        return connection.getInputStream();
    }

    @Nullable
    @MainThread
    public static DisplayImageAction<?> getDisplayImageAction(@Nullable final View view) {
        if (view != null) {
            final Drawable drawable = getDrawable(view);
            if (drawable instanceof PlaceholderDrawable) {
                return ((PlaceholderDrawable) drawable).getAction();
            }
        }
        return null;
    }

    public static void setDrawable(@NonNull final Drawable drawable, @NonNull final View view) {
        if (view instanceof ImageView) {
            ((ImageView) view).setImageDrawable(drawable);
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                view.setBackground(drawable);
            } else {
                view.setBackgroundDrawable(drawable);
            }
        }
    }

    public static void setBitmap(@NonNull final Resources resources, @NonNull final Bitmap bitmap,
            @NonNull final View view) {
        if (view instanceof ImageView) {
            ((ImageView) view).setImageBitmap(bitmap);
        } else {
            final Drawable drawable = new BitmapDrawable(resources, bitmap);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                view.setBackground(drawable);
            } else {
                view.setBackgroundDrawable(drawable);
            }
        }
    }

    @Nullable
    public static Drawable getDrawable(@NonNull final View view) {
        if (view instanceof ImageView) {
            return ((ImageView) view).getDrawable();
        } else {
            return view.getBackground();
        }
    }

    public static int getLoadPoolSize() {
        return Math.min(Runtime.getRuntime().availableProcessors(), MAX_POOL_SIZE);
    }

    public static int getCachePoolSize() {
        final int size = getLoadPoolSize() / 2;
        if (size < MIN_POOL_SIZE) {
            return MIN_POOL_SIZE;
        } else {
            return size;
        }
    }

    public static boolean isUriLocal(@NonNull final Uri uri) {
        return isUriSchemeLocal(uri.getScheme());
    }

    public static boolean isUriLocal(@NonNull final String uri) {
        final int ssi = uri.indexOf(':');
        return ssi != -1 && isUriSchemeLocal(uri.substring(0, ssi));

    }

    private static boolean isUriSchemeLocal(@NonNull final String scheme) {
        return ContentResolver.SCHEME_FILE.equals(scheme) || ContentResolver.SCHEME_CONTENT.equals(scheme) ||
                ContentResolver.SCHEME_ANDROID_RESOURCE.equals(scheme);
    }

    public static int getExifRotation(@NonNull final Context context, @NonNull final Uri uri) {
        InputStream inputStream = null;
        try {
            inputStream = context.getContentResolver().openInputStream(uri);
            if (inputStream != null) {
                return getExifRotation(new ExifInterface(inputStream));
            } else {
                return 0;
            }
        } catch (final IOException e) {
            return 0;
        } finally {
            close(inputStream);
        }
    }

    public static int getExifRotation(@NonNull final File file) {
        try {
            return getExifRotation(new ExifInterface(file.getAbsolutePath()));
        } catch (final IOException e) {
            return 0;
        }
    }

    public static int getExifRotation(@NonNull final byte[] bytes) {
        try {
            return getExifRotation(new ExifInterface(new ByteArrayInputStream(bytes)));
        } catch (final IOException e) {
            return 0;
        }
    }

    public static int getExifRotation(@NonNull final ExifInterface exifInterface) {
        switch (exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
            case ExifInterface.ORIENTATION_ROTATE_90:
                return 90;
            case ExifInterface.ORIENTATION_ROTATE_180:
                return 180;
            case ExifInterface.ORIENTATION_ROTATE_270:
                return 270;
            default:
                return 0;
        }
    }

    @NonNull
    public static Bitmap rotateAndRecycle(@NonNull final Bitmap bitmap, final int rotation) {
        final Bitmap rotated = ImageUtils.rotate(bitmap, rotation);
        if (bitmap != rotated) {
            bitmap.recycle();
        }
        return rotated;
    }

    @NonNull
    public static <T> T requireNonNull(@Nullable final T value) {
        if (value == null) {
            throw new NullPointerException();
        }
        return value;
    }
}

File: src/main/java/com/budiyev/android/imageloader/InternalUtils.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 Yuriy Budiyev [yuriy.budiyev@yandex.ru]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.budiyev.android.imageloader

import android.content.ContentResolver
import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.support.annotation.MainThread
import android.support.media.ExifInterface
import android.view.View
import android.widget.ImageView
import java.io.ByteArrayInputStream
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.URL

internal object InternalUtils {
    private val CONNECT_TIMEOUT = 10000
    private val MAX_POOL_SIZE = 4
    private val MIN_POOL_SIZE = 1
    private val URI_SCHEME_HTTP = "http"
    private val URI_SCHEME_HTTPS = "https"
    private val URI_SCHEME_FTP = "ftp"

    val loadPoolSize: Int
        get() = Math.min(Runtime.getRuntime().availableProcessors(), MAX_POOL_SIZE)

    val cachePoolSize: Int
        get() {
            val size = loadPoolSize / 2
            return if (size < MIN_POOL_SIZE) {
                MIN_POOL_SIZE
            } else {
                size
            }
        }

    fun invalidate(memoryCache: ImageCache?, storageCache: ImageCache?, descriptor: DataDescriptor<*>) {
        val key = descriptor.key ?: return
        memoryCache?.remove(key)
        storageCache?.remove(key)
    }

    fun buildFullKey(base: String?, requiredSize: Size?, transformation: BitmapTransformation?): String? {
        if (base == null) {
            return null
        }
        if (requiredSize == null && transformation == null) {
            return base
        }
        val sb = StringBuilder(base)
        if (requiredSize != null) {
            sb.append("_required_size_").append(requiredSize.width).append("x").append(requiredSize.height)
        }
        if (transformation != null) {
            sb.append(transformation.key)
        }
        return sb.toString()
    }

    fun close(closeable: Closeable?) {
        if (closeable == null) {
            return
        }
        try {
            closeable.close()
        } catch (ignored: IOException) {
        }

    }

    @Throws(IOException::class)
    fun getDataStreamFromUri(context: Context, uri: Uri): InputStream? {
        val scheme = uri.scheme
        if (URI_SCHEME_HTTP.equals(scheme, ignoreCase = true) || URI_SCHEME_HTTPS.equals(scheme,
                        ignoreCase = true) || URI_SCHEME_FTP.equals(scheme, ignoreCase = true)) {
            val connection = URL(uri.toString()).openConnection()
            connection.connectTimeout = CONNECT_TIMEOUT
            return connection.getInputStream()
        } else {
            return context.contentResolver.openInputStream(uri)
        }
    }

    @Throws(IOException::class)
    fun getDataStreamFromUrl(url: String): InputStream? {
        val connection = URL(url).openConnection()
        connection.connectTimeout = CONNECT_TIMEOUT
        return connection.getInputStream()
    }

    @MainThread
    fun getDisplayImageAction(view: View?): DisplayImageAction<*>? {
        if (view != null) {
            val drawable = getDrawable(view)
            if (drawable is PlaceholderDrawable) {
                return drawable.action
            }
        }
        return null
    }

    fun setDrawable(drawable: Drawable, view: View) {
        if (view is ImageView) {
            view.setImageDrawable(drawable)
        } else {
            view.background = drawable
        }
    }

    fun setBitmap(resources: Resources, bitmap: Bitmap, view: View) {
        if (view is ImageView) {
            view.setImageBitmap(bitmap)
        } else {
            view.background = BitmapDrawable(resources, bitmap)
        }
    }

    private fun getDrawable(view: View): Drawable? {
        return if (view is ImageView) {
            view.drawable
        } else {
            view.background
        }
    }

    fun isUriLocal(uri: Uri): Boolean {
        return isUriSchemeLocal(uri.scheme)
    }

    fun isUriLocal(uri: String): Boolean {
        val ssi = uri.indexOf(':')
        return ssi != -1 && isUriSchemeLocal(uri.substring(0, ssi))

    }

    private fun isUriSchemeLocal(scheme: String): Boolean {
        return ContentResolver.SCHEME_FILE == scheme || ContentResolver.SCHEME_CONTENT == scheme || ContentResolver.SCHEME_ANDROID_RESOURCE == scheme
    }

    fun getExifRotation(context: Context, uri: Uri): Int {
        var inputStream: InputStream? = null
        try {
            inputStream = context.contentResolver.openInputStream(uri)
            return if (inputStream != null) {
                getExifRotation(ExifInterface(inputStream))
            } else {
                0
            }
        } catch (e: IOException) {
            return 0
        } finally {
            close(inputStream)
        }
    }

    fun getExifRotation(file: File): Int {
        try {
            return getExifRotation(ExifInterface(file.absolutePath))
        } catch (e: IOException) {
            return 0
        }

    }

    fun getExifRotation(bytes: ByteArray): Int {
        try {
            return getExifRotation(ExifInterface(ByteArrayInputStream(bytes)))
        } catch (e: IOException) {
            return 0
        }

    }

    fun getExifRotation(exifInterface: ExifInterface): Int {
        when (exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
            ExifInterface.ORIENTATION_ROTATE_90 -> return 90
            ExifInterface.ORIENTATION_ROTATE_180 -> return 180
            ExifInterface.ORIENTATION_ROTATE_270 -> return 270
            else -> return 0
        }
    }

    fun rotateAndRecycle(bitmap: Bitmap, rotation: Int): Bitmap {
        val rotated = ImageUtils.rotate(bitmap, rotation.toFloat())
        if (bitmap != rotated) {
            bitmap.recycle()
        }
        return rotated
    }

    fun <T> requireNonNull(value: T?): T {
        if (value == null) {
            throw NullPointerException()
        }
        return value
    }
}
